File: src/lib/sekiban-go/mv/query_port.go
package mv

import (
	"encoding/json"
	"fmt"

	"github.com/J-Tech-Japan/sekiban-go/wasm"
)

// HostBackedQueryPort routes QueryRows / QuerySingleRow through the `env.mv_host_query_rows`
// host import provided by the wasmserver. The sample projector `ClassRoomEnrollmentMvV1`
// does not query mid-apply so this type exists primarily so future projectors can drop it
// in without re-declaring the import.
type HostBackedQueryPort struct{}

func NewHostBackedQueryPort() HostBackedQueryPort { return HostBackedQueryPort{} }

// QueryRows panics if the host returns a `{"error":...}` envelope. That's the contract the
// C# host relies on so a failing mid-apply query aborts the MV apply and lets Orleans retry —
// silently dropping the error would produce stale projections. Projectors that prefer soft
// handling should call invokeSoft directly.
func (HostBackedQueryPort) QueryRows(sql string, params []MvParam) []MvQueryRowDto {
	result, err := invoke(sql, params, 0x7fffffff)
	if err != nil {
		panic(fmt.Errorf("mv_host_query_rows failed: %w", err))
	}
	if result == nil {
		return nil
	}
	return result.Rows
}

func (HostBackedQueryPort) QuerySingleRow(sql string, params []MvParam) *MvQueryRowDto {
	result, err := invoke(sql, params, 1)
	if err != nil {
		panic(fmt.Errorf("mv_host_query_rows failed: %w", err))
	}
	if result == nil || len(result.Rows) == 0 {
		return nil
	}
	first := result.Rows[0]
	return &first
}

func invoke(sql string, params []MvParam, rowLimit int32) (*MvQueryResultDto, error) {
	sqlBytes := []byte(sql)
	var paramsBytes []byte
	if len(params) > 0 {
		data, err := json.Marshal(params)
		if err != nil {
			return nil, fmt.Errorf("marshal params: %w", err)
		}
		paramsBytes = data
	}

	sqlPtr := bytesPointer(sqlBytes)
	paramsPtr := bytesPointer(paramsBytes)

	packed := hostMvHostQueryRows(sqlPtr, uint32(len(sqlBytes)), paramsPtr, uint32(len(paramsBytes)), rowLimit)
	ptr, length := wasm.UnpackPtrLen(packed)
	if ptr == 0 || length == 0 {
		return nil, nil
	}
	text := wasm.ReadString(ptr, length)
	// `wasm.Dealloc` is intentionally a no-op (TinyGo reactor mode is sensitive to manual
	// free). The allocation registry is cleared between host-driven MV export invocations by
	// `wasm.ResetAllocations()` in ExportHelpers, so repeated mid-apply queries do not
	// accumulate across calls.
	wasm.Dealloc(ptr, length)

	var errEnvelope struct {
		Error string `json:"error,omitempty"`
	}
	if err := json.Unmarshal([]byte(text), &errEnvelope); err == nil && errEnvelope.Error != "" {
		return nil, fmt.Errorf("%s", errEnvelope.Error)
	}
	var result MvQueryResultDto
	if err := json.Unmarshal([]byte(text), &result); err != nil {
		return nil, fmt.Errorf("parse mv_host_query_rows response: %w", err)
	}
	return &result, nil
}

func bytesPointer(b []byte) uint32 {
	if len(b) == 0 {
		return 0
	}
	return wasm.BytesPointer(b)
}

File: src/wasm-projectors/swift/sekiban-mv/Sources/SekibanMv/QueryPort.swift
import Foundation
import SekibanWasm

// Host import from the wasmserver. Signature: (sql_ptr, sql_len, params_ptr, params_len,
// row_limit) -> packed i64 pointing at a UTF-8 JSON MvQueryResultDto (or `{"error":"..."}`).
//
// `@_silgen_name` alone would emit Swift's native calling convention (7 i32 params including
// the implicit error slot) which wasm-ld records as the import signature. wasmtime then refuses
// to link because the host declares `(i32 i32 i32 i32 i32) -> i64`. `@_extern(c, ...)` forces
// the C ABI so the signatures match.
@_extern(wasm, module: "env", name: "mv_host_query_rows")
@_extern(c)
func host_mv_host_query_rows(
    _ sqlPtr: Int32, _ sqlLen: Int32,
    _ paramsPtr: Int32, _ paramsLen: Int32,
    _ rowLimit: Int32
) -> Int64

/// Default query port that routes through the `env.mv_host_query_rows` host import.
/// The sample's `ClassRoomEnrollmentMvV1` doesn't query mid-apply so this exists primarily to
/// let future projectors call into the host without each one having to re-declare the `extern`.
public struct HostBackedMvQueryPort: MvQueryPort {
    public init() {}

    public func queryRows(_ sql: String, _ params: [MvParam]) -> [MvQueryRowDto] {
        invoke(sql: sql, params: params, rowLimit: Int32.max)?.rows ?? []
    }

    public func querySingleRow(_ sql: String, _ params: [MvParam]) -> MvQueryRowDto? {
        invoke(sql: sql, params: params, rowLimit: 1)?.rows.first
    }

    private func invoke(sql: String, params: [MvParam], rowLimit: Int32) -> MvQueryResultDto? {
        let sqlBytes = Array(sql.utf8)
        let paramsJson: [UInt8]
        if params.isEmpty {
            paramsJson = []
        } else if let data = try? JSONEncoder().encode(params) {
            paramsJson = Array(data)
        } else {
            return nil
        }

        // Temporarily hold the bytes in arrays so their pointers stay valid across the call.
        return sqlBytes.withUnsafeBufferPointer { sqlBuf in
            paramsJson.withUnsafeBufferPointer { paramsBuf in
                let sqlPtr: Int32 = Int32(truncatingIfNeeded:
                    UInt32(UInt(bitPattern: UnsafeRawPointer(sqlBuf.baseAddress ?? UnsafePointer<UInt8>(bitPattern: 0)!))))
                let sqlLen = Int32(sqlBytes.count)
                let paramsPtr: Int32 = paramsJson.isEmpty ? 0 : Int32(truncatingIfNeeded:
                    UInt32(UInt(bitPattern: UnsafeRawPointer(paramsBuf.baseAddress ?? UnsafePointer<UInt8>(bitPattern: 0)!))))
                let paramsLen = Int32(paramsJson.count)

                let packed = host_mv_host_query_rows(sqlPtr, sqlLen, paramsPtr, paramsLen, rowLimit)
                let (ptr, len) = unpackPtrLen(packed)
                if ptr == 0 || len == 0 { return nil }
                let text = readString(ptr: ptr, len: len)
                // Host allocated the buffer via our `alloc` export, so free it now.
                let raw = UnsafeMutableRawPointer(bitPattern: Int(UInt32(bitPattern: ptr)))
                free(raw)
                if let errData = text.data(using: .utf8),
                   let obj = try? JSONSerialization.jsonObject(with: errData) as? [String: Any],
                   obj["error"] != nil {
                    return nil
                }
                guard let data = text.data(using: .utf8) else { return nil }
                return try? JSONDecoder().decode(MvQueryResultDto.self, from: data)
            }
        }
    }
}
